gnomad_qc/v4/resources/basics.py

```diff
@@ -146,35 +140,110 @@ def _remove_ukbb_dup_by_index(
 )
 
 
-# UKBB data resources
-def _ukbb_root_path() -> str:
+# UKB data resources
+def _ukb_root_path() -> str:
     """
-    Retrieve the path to the UKBB data directory.
+    Retrieve the path to the UKB data directory.
 
-    :return: String representation of the path to the UKBB data directory
+    :return: String representation of the path to the UKB data directory
     """
     return "gs://gnomad/v4.0/ukbb"
 
 
-# List of samples to exclude from QC due to withdrawn consents
-# This is the file with the most up to date sample withdrawals we were sent. File downloaded on 8/16/21
-ukbb_excluded_samples_path = f"{_ukbb_root_path()}/w26041_20210809.csv"
+# List of samples to exclude from QC due to withdrawn consents for application 26041 on 08/09/2021 and application 31063 on 02/22/2022, originally imported at csvs and keyed by their respective eids.
+ukb_excluded = VersionedTableResource(
+    default_version="31063_20220222",
+    versions={
+        "26041_20210809": TableResource(path=f"{_ukb_root_path()}/w26041_20210809.ht"),
+        "31063_20220222": TableResource(
+            path=f"{_ukb_root_path()}/ukb31063.withdrawn_participants.20220222.ht"
+        ),
+    },
+)
+
+# UKB map of exome IDs to array sample IDs (application ID: 26041)
+ukb_array_sample_map = TableResource(f"{_ukb_root_path()}/array_sample_map_freeze_7.ht")
 
-# UKBB map of exome IDs to array sample IDs (application ID: 26041)
-ukbb_array_sample_map = TableResource(
-    f"{_ukbb_root_path()}/array_sample_map_freeze_7.ht"
+# UKB full mapping file of sample ID and application IDs 26041, 31063, and 48511
+ukb_application_map = TableResource(
+    f"{_ukb_root_path()}/ukbb_application_id_mappings.ht"
 )
 
+
+def generate_ukb_application_map() -> None:
+    """
+    Generate the mapping file of application IDs 26041, 31063, and 48511, and add in annotatation 'withdraw', which is set to True if the UKB sample needs to be withdrawn.
+
+    :return: None
+    """
+    # The array sample map has the most complete list of 's' and 'eid_26041' for UKB samples, so use this file rather than the UKB 'meta.ht' (which already has some samples withdrawn from it)
+    ukb_array_sample_map = ukb_array_sample_map.ht()
+    ukb_array_sample_map = ukb_array_sample_map.select(
+        eid_26041=ukb_array_sample_map.ukbb_app_26041_id
+    )
+
+    # Add in bridge mapping from eid_26041 to eid_31063
+    ukb_map = ukb_array_sample_map.key_by("eid_26041")
```

Comment:
This is because the resource you are using has the same name, also we typically add `_ht` for a HT. I also prefer changing `ukb_map` -> `ukb_map_ht`, but not needed.

```suggestion
    ukb_array_sample_map_ht = ukb_array_sample_map.ht()
    ukb_array_sample_map_ht = ukb_array_sample_map_ht.select(
        eid_26041=ukb_array_sample_map_ht.ukbb_app_26041_id
    )

    # Add in bridge mapping from eid_26041 to eid_31063
    ukb_map = ukb_array_sample_map.key_by("eid_26041")
```
```suggestion
    ukb_array_sample_map = ukb_array_sample_map.ht()
    ukb_array_sample_map = ukb_array_sample_map.select(
        eid_26041=ukb_array_sample_map.ukbb_app_26041_id
    )

    # Add in bridge mapping from eid_26041 to eid_31063
    ukb_map = ukb_array_sample_map.key_by("eid_26041")
```